modules/lib/src/main/resources/assets/js/v6/features/layout/AppShell/Sidebar.tsx

```diff
@@ -0,0 +1,242 @@
+
+type WidgetHandler = (widgetId: string) => void;
+
+type Props = {
+  name?: string;
+  widgets?: Widget[];
+  activeWidgetId?: string;
+  handler?: WidgetHandler;
+};
+
+const WidgetButton = ({
+  widget,
+  activeWidgetId,
+  handler,
+}: Pick<Props, "activeWidgetId" | "handler"> & { widget: Widget }): React.ReactElement => {
+  const [widgetKey, widgetDisplayName, widgetIconUrl, widgetFullIconUrl] = [
+    widget.getWidgetDescriptorKey().toString(),
+    widget.getDisplayName(),
+    widget.getIconUrl(),
+    widget.getFullIconUrl(),
+  ];
+  const baseIcon = CircleQuestionMark;
+  const isWidgetActive = widgetKey === activeWidgetId;
+  const icon = SidebarElement.isMainWidget(widget) ? Pen : baseIcon;
+
+  if (widgetIconUrl) {
+    return (
+      <Tooltip value={widgetDisplayName} side="right">
+        <Button
+          className={cn(
+            "px-[13px]",
+            isWidgetActive
+              ? "bg-surface-primary-selected hover:bg-surface-primary-selected-hover"
+              : ""
+          )}
+          size="lg"
+          aria-label={widget.getDisplayName()}
+          onClick={() => handler(widgetKey)}
+        >
+          <img
+            class={cn(
+              "w-[20px] invert-100 dark:invert-0",
+              isWidgetActive ? "invert-0" : ""
+            )}
+            src={widgetFullIconUrl}
+            alt={widgetDisplayName}
+          />
+        </Button>
+      </Tooltip>
+    );
+  }
+
+
+  return (
+    <Tooltip value={widgetDisplayName} side="right">
+      <IconButton
+        className={
+          isWidgetActive
+            ? "bg-surface-primary-selected hover:bg-surface-primary-selected-hover text-white"
+            : "text-black dark:text-white"
+        }
+        size="lg"
+        icon={icon}
+        aria-label={widget.getDisplayName()}
+        onClick={() => handler(widgetKey)}
+      />
+    </Tooltip>
+  );
+};
+
+const SidebarComponent = ({
+  name = Store.instance().get("application").getName(),
+  widgets = [],
+  activeWidgetId = "",
+  handler = () => {},
+}: Props): React.ReactElement => {
+  const project = ProjectContext.get().getProject();
+  const version = "v." + CONFIG.getString("appVersion");
+  const mainWidgets = widgets.slice(0, -1);
+  const lastWidget = widgets.at(-1);
+
+  const SidebarHeader = () => (
+    <>
+      <div class="w-[32px] h-[32px]">
+        <ProjectIcon
+          projectName={project.getName()}
+          language={project.getLanguage()}
+          hasIcon={!!project.getIcon()}
+        />
+      </div>
+      <p
+        title={name}
+        class="[writing-mode:vertical-lr] text-nowrap text-base font-semibold"
+      >
+        {name}
+      </p>
+    </>
+  );
+
+  const SidebarWidgets = () => (
+    <div className="flex flex-col gap-2 items-center">
+      {mainWidgets.map((widget) => (
+        <WidgetButton
+          widget={widget}
+          activeWidgetId={activeWidgetId}
+          handler={handler}
+        />
```

Comment:
Must pass `key` to `<WidgetButton>` to avoid re-rendering. Can probably use `String(widget.getWidgetDescriptorKey())` for that.